### ISSUE TITLE
Enigma: error importing GPG key, no trustdb.gpg

### ISSUE THREAD

**Issue by jimc-leones** [roundcube/roundcubemail#5526]
Versions: roundcubemail-1.2.2
gpg2-2.0.24
apache2-mod_php5-5.5.14
apache2-2.4.16
OpenSuSE "Leap" 42.1

Enigma configuration: none, take all defaults.  
Starting with no keyring at all, I import a keypair, and 
/var/log/roundcube/errors reports: 
[18-Nov-2016 14:44:09 -0800]: <pgdg1rgb> PHP Error: 
    Enigma plugin: Unknown error importing GPG key. 
    Please use the 'debug' option when creating the Crypt_GPG object, 
    and file a bug report at 
    http://pear.php.net/bugs/report.php?package=Crypt_GPG in 
    /srv/www/roundcubemail/plugins/enigma/lib/enigma_engine.php 
    on line 1079 
    (POST /kerberos/roundcube/?_task=settings&_action=plugin.enigmakeys\
    &_a=import&_unlock=loading1479509048799)

I've attached the debug output.  Crypt_GPG is not at fault; it's an
Enigma (or GPG?) issue.  The key item (IMHO) is:
gpg: Fatal: can't open 
`/home/httpd/htdocs/kerberos/roundcube/plugins/enigma/home/jimc/trustdb.gpg': 
No such file or directory

Which is no lie.  Even so the keys were added to the keyrings.  An
empty file is not acceptable as trustdb.gpg.  To create one I did:
su wwwrun -s /bin/sh -c "gpg --no-default-keyring --no-options \
    --no-permission-warning --trust-model pgp --homedir \
    '/home/httpd/htdocs/kerberos/roundcube/plugins/enigma/home/jimc' \
    --update-trustdb"

Note --trust-model 'pgp', not 'always'.  trustdb.gpg was created, and 
now I can import other keys, or delete keys, without getting the 
error message.  

As I see it, with --trust-model 'always', gpg should not be messing 
with trustdb.gpg at all, but maybe this version of gpg assumes that
it's important to assess trust even when told not to.  

I kind of assumed that Enigma was going to use --trust-model 'pgp' 
by default.  But I notice in /srv/www/roundcubemail/plugins/enigma/README
that there's a ToDo item for "Mark keys as trusted/untrasted, 
display appropriate message in verify/decrypt status".  I've written 
something on signed mail, see 
http://www.jfcarter.net/~jimc/documents/mail-signed-1611.html
Particularly see the section on Key Management: for the recipient
to resist a signed phishing message he has to know that the key by
which it was signed is not trusted.  It would be really helpful if
signatures on keys could be honored and the message signature's 
trust status announced, even if Enigma's GUI is not extensive 
enough to manage the Web of Trust itself.

**Comment by jimc-leones** [comment]
Here's the Crypt_GPG debug output.
[rc-log.txt](https://github.com/roundcube/roundcubemail/files/601799/rc-log.txt)

**Comment by alecpl** [comment]
This looks to me as GnuPG issue fixed here: https://git.gnupg.org/cgi-bin/gitweb.cgi?p=gnupg.git;a=commitdiff;h=07006c9916ea194ce6047d252421c08489068c4c i.e. in  2.0.27. The version you're using is pretty outdated. I don't think we should spend time on finding workarounds.

**Comment by alecpl** [comment]
For gnupg-2.1 you need Roundcube from git-master and Crypt_GPG-1.6.x

**Comment by jimc-leones** [comment]
Thanks for the quick work making Crypt_GPG-1.6.0b2 .  I did some 
testing.

Starting with roundcubemail-1.2.2, php5-pear-Crypt_GPG-1.4.0-2.1.noarch,
gpg2-2.0.24-5.1.x86_64, I upgraded things step by step.  

I upgraded to gpg2-2.0.26-2.6.1.x86_64 from OpenSuSE-13.2.  gpg2-lang 
also has to be upgraded (or removed).  No complaints about other 
libraries.  The SuSE distro manager says that this version includes
the commit that fixes the problem with opening trustdb.gpg wrongly. 

Starting with no keyring, I uploaded keys, and sent some
messages that gave trouble to another mail crypto package.  No errors
in /var/log/roundcube/errors, messages were signed, and received 
signatures were verified.  

I removed php5-pear-Crypt_GPG-1.4.0-2.1.noarch and did
pear install Crypt_GPG-1.6.0b2 (successful).  I cloned 
https://github.com/roundcube/roundcubemail (master) and replaced 
only the Enigma plugin, hiding the version from 1.2.2 and replacing 
with the version from the git clone.  config.inc did not exist 
(taking all defaults).  .../enigma/home changed to mode 700 wwwrun:www.
Starting with no keyring.  Imported the keys,  They look correct, and
no error messages.  I sent a simple message.  The pass phrase 
dialog is repeated multiple times, i.e. the pass phrase does not 
unlock the key.

I tried gpg by hand:
    /usr/bin/gpg --no-secmem-warning --no-tty --no-default-keyring \
    --no-options --no-permission-warning --exit-on-status-write-error \
    --trust-model always --homedir \
    '/home/httpd/htdocs/kerberos/roundcube/plugins/enigma/home/jimc' \
    --local-user jimc@jfcarter.net --clearsign gorf.txt 
It did sign the file, so there is nothing wrong with secring.gpg .
(This is Crypt_GPG-1.6.0b2 with gpg2-2.0.26 .)

I then upgraded to gpg2-2.1.16-198.1.x86_64 from the OpenSuSE 
"Leap" 42.1 security:privacy sub-repo (42.1 is the distro version
I'm running.)  I sent a simple message, which was sent OK using the
keyring created by 2.0.26 and using the passphrase it had acquired
earlier.  But reading the received message: "Unverified signature,
internal error".  In /var/log/roundcube/errors: "Enigma plugin: 
Unknown error validating signature details."  

I didn't put too much time into debugging this; I shredded the 
keyrings and re-imported the keys.  They are now stored in pubring.kbx
(mode 644), a "keybox".  Reading the same message that formerly got 
an unverified signature: now verified.  I sent more of my test 
messages and they were sent and verified OK with no errors reported.

I'm not sure how backward compatible things are supposed to be, but
it looks like we have a consistent set of packages that works 
error-free, within the limits of my testing.  I'm going to suggest
to the SuSE distro manager that he coordinate with you to pick the
versions, and then upgrade Roundcube (and Enigma in it).
